ops/testing.py

```diff
@@ -2216,17 +2325,34 @@ def _storage_remove(self, storage_id: str):
         index = int(index)
         self._storage_list[name].pop(index, None)
 
-    def action_get(self):  # type:ignore
-        raise NotImplementedError(self.action_get)  # type:ignore
-
-    def action_set(self, results):  # type:ignore
-        raise NotImplementedError(self.action_set)  # type:ignore
-
-    def action_log(self, message):  # type:ignore
-        raise NotImplementedError(self.action_log)  # type:ignore
-
-    def action_fail(self, message=''):  # type:ignore
-        raise NotImplementedError(self.action_fail)  # type:ignore
+    def action_get(self) -> Dict[str, Any]:
+        action_name = os.environ["JUJU_ACTION_NAME"]
+        params: Dict[str, Any] = {}
+        for name, action_meta in self._meta.actions[action_name].parameters.items():
+            if "default" in action_meta:
+                params[name] = action_meta["default"]
+        params.update(self._operation.parameters)
+        return params
+
+    def action_set(self, results: Dict[str, Any]):
+        for key in "stdout", "stderr", "stdout-encoding", "stderr-encoding":
```

Comment:
Huh, I've never seen the implicit-comma-builds-tuple used in this context. I think it's slightly clearer to be explicit:

```suggestion
        for key in ["stdout", "stderr", "stdout-encoding", "stderr-encoding"]:
```

Reply:
> Huh, I've never seen the implicit-comma-builds-tuple used in this context. I think it's slightly clearer to be explicit:

Ok (my memory was that black trimmed them in that case, but (a) autopep8 doesn't, and (b) a quick test seems to indicate it doesn't). However, a tuple not a list:

```suggestion
        for key in ("stdout", "stderr", "stdout-encoding", "stderr-encoding"):
```

Reply:
Hmm, in older versions of Python a tuple was fastest (and to my eyes looks cleanest - a frozenset is arguably the most correct but is the ugliest and slowest). However, it seems like in Python 3.8+ a list is faster (although a bit heavier in terms of memory).

I still like the look of a tuple most, so will go with that, but re-open this if you disagree :)

Reply:
I *definitely* don't think we need to worry about speed here. I'm fine with either a tuple or a list, so tuple is fine!

However, just for the record, here's how I usually think about the distinction: it's less about immutable vs mutable, and more about: tuples are "records with positional fields", and lists are "a list of same-typed items". So I'd use `(name, email)` but `[name1, name2, ...]`. There's a fair bit of support for this take [on StackOverflow](https://stackoverflow.com/a/626871/68707) and [in the official tutorial](https://docs.python.org/3/tutorial/datastructures.html#tuples-and-sequences), which says:

> Though tuples may seem similar to lists, they are often used in different situations and for different purposes. Tuples are [immutable](https://docs.python.org/3/glossary.html#term-immutable), and usually contain a heterogeneous sequence of elements that are accessed via unpacking (see later in this section) or indexing (or even by attribute in the case of [namedtuples](https://docs.python.org/3/library/collections.html#collections.namedtuple)). Lists are [mutable](https://docs.python.org/3/glossary.html#term-mutable), and their elements are usually homogeneous and are accessed by iterating over the list.